### ISSUE TITLE
Questions about the code.

### ISSUE THREAD

**Issue by SuleBai** [leaves162/CLIPtrase#2]
Hi,

Thank you for your impressive work! I have a few questions regarding the code implementation.

After reviewing the code, I noticed that in `attn2.py`, the attention weight for multiplying the value embeddings is still calculated using the inner product rather than the cosine similarity mentioned in the paper. I understand that the cosine similarity one is used for the clustering operation later, but I’m curious about the decision to use inner-product attention weights when multiplying the value embeddings.

Additionally, in `Table4`, rows 3 and 4, if I understand correctly, these results were obtained without applying the clustering operation, and the cosine similarity attention weight performs slightly better than the inner product attention weight. If that’s the case, I’m confused why cosine similarity wasn’t used for multiplying the value embeddings as well.
<img width="1015" alt="image" src="https://github.com/user-attachments/assets/9edaa6aa-d067-447d-ab13-7dfa97e4c35b">

Thank you for your time, and I look forward to your clarification!

**Comment by leaves162** [comment]
Thank you for your attention to our work!

Both inner product and cosine can play the role of self-correlation (The difference is whether there is norm). When the result of cosine is directly used as weights for softmax, it is necessary to design an additional temperature coefficient. The purpose of using cosine is to use its limited value range to facilitate the subsequent selection of hyperparameters in clustering. In fact, we are still optimizing this work, and we will update it later soon.

Hope our reply can help you.

**Comment by SuleBai** [comment]
Hi, thanks for your timely reply!

You mentioned that when the cosine similarity result is directly used as weights for softmax, an additional temperature coefficient is necessary.

Could you please explain how you designed the temperature coefficient to achieve the results in Row 4 of Table 4? I attempted to reproduce this result by directly using the `attention_output_weights_ori` below as the attention weight, but unfortunately, my results were quite poor.

```
for b in range(q.shape[0]):
    qq_cos = torch.cosine_similarity(
        q[b].unsqueeze(dim=1), q[b].unsqueeze(dim=0), dim=-1
    ) # l,l
    kk_cos = torch.cosine_similarity(
        k[b].unsqueeze(dim=1), k[b].unsqueeze(dim=0), dim=-1
    ) # l,l
    vv_cos = torch.cosine_similarity(
        v[b].unsqueeze(dim=1), v[b].unsqueeze(dim=0), dim=-1
    ) # l,l
    attn_output_weights_ori.append((qq_cos+kk_cos+vv_cos)/3.0)
attn_output_weights_ori = torch.stack(attn_output_weights_ori,dim=0) # b,l,l
```

Thank you for your assistance!

**Comment by leaves162** [comment]
Sure, you can add norm based on the implementation of inner-product to transform it into cosine, and multiply about 7 (I remembered it should be this, a little different between datasets), and remember to delete the scaling, other implementation keeps unchanged. 

As for directly using the code of generating attention weights to replace the inner product operation, we found the mean of qq, kk, vv first and then doing softmax degrade the performance, where we think this is because "mean" confuses the learning focus of different branches.
